src/Http/Http.Abstractions/src/Extensions/UseMiddlewareExtensions.cs

```diff
@@ -215,13 +216,58 @@ public RequestDelegate CreateMiddleware(RequestDelegate next)
             methodArguments[0] = context;
             for (var i = 1; i < parameters.Length; i++)
             {
-                methodArguments[i] = GetService(serviceProvider, parameters[i].ParameterType, methodInfo.DeclaringType!);
+                var parameter = parameters[i];
+
+                var hasServiceKey = TryGetServiceKey(parameter, out object? key);
```

Comment:
This is going to happen on every single request, probably worth moving outside of the factory method since the data is the same for every single call.

Reply:
Hi Brennan, 

Thanks for your valuable feedback.

I've made some changes to reflect the feedback. I now precompute and cache the results for each parameter.

Reply:
@BrennanConroy, I could eliminate the condition check (hasServiceKey ? GetKeyedService(...) : GetService(...)) as wel by compiling the method calls into delegates, and calling to the delegate directly within the factory. We'd probably be talking about microseconds though. The code might become slightly less readable though.

What's your opinion on the matter?